src/Course/State.hs

```diff
@@ -113,13 +92,21 @@ instance Applicative (State s) where
 --
 -- >>> let modify f = State (\s -> ((), f s)) in runState (modify (+1) >>= \() -> modify (*2)) 7
 -- ((),16)
+
+-- Earlier attempt: this passes tests but causes later exercises to fail, I'm sure it's stupid
+-- but I'll leave it here to come back and understand why
+--
+-- instance Monad (State s) where
+--   (=<<) :: (a -> State s b) -> State s a -> State s b
+--   (=<<) f2 s1 = State (\x -> (e1 x, e2 x))
+--     where
+--       s2 x = f2 (eval s1 x) -- 2nd state to apply is a function of the first state's evaluation
+--       e1 x = eval (s2 x) x
+--       e2 x = exec (s2 x) (exec s1 x)
+
 instance Monad (State s) where
-  (=<<) ::
-    (a -> State s b)
-    -> State s a
-    -> State s b
-  (=<<) =
-    error "todo: Course.State (=<<)#instance (State s)"
+  (=<<) :: (a -> State s b) -> State s a -> State s b
+  (=<<) f state1 = State $ \s -> let (val1, s1) = runState state1 s in runState (f val1) s1 -- just copied this from the Internet
```

Comment:
One liners are a bit tricky to understand... but sorry, that's obvious to you... you're just trying to get it work, as you say.

My version might be worth looking at, might help you understand it a bit more...

https://github.com/JulianLeviston/fp-course/pull/7/files#diff-ad52b034ecdacc8f3d5b1e973ac02c79R136

I think the key point is that in monad, we're taking each result and doing the next thing... sequencing, kinda.